File: rust/muse/src/algorithms/search/binary_search.rs
use std::cmp::Ordering;

pub fn find(arr: &[i32], key: i32) -> usize {
    let mut lo = 0;
    let mut hi = arr.len();

    while lo < hi {
        let mid = lo + ((hi - lo) >> 1);
        match key.cmp(&arr[mid]) {
            Ordering::Less => {
                hi = mid;
            }
            Ordering::Greater => {
                lo = mid + 1;
            }
            Ordering::Equal => {
                return mid;
            }
        };
    }

    arr.len()
}

#[cfg(test)]
mod tests;

File: python/muse/algorithms/search/binary_search.py
def find(arr: list, key: object) -> int:
    lo = 0
    hi = len(arr)

    while lo < hi:
        mid = lo + ((hi - lo) >> 1)
        if key < arr[mid]:
            hi = mid
        elif key > arr[mid]:
            lo = mid + 1
        else:
            return mid

    return len(arr)
